src/slice.rs

```diff
@@ -127,6 +127,87 @@ impl<'a> RopeSlice<'a> {
         })
     }
 
+    pub(crate) fn new_with_byte_range(
+        node: &'a Arc<Node>,
+        start: usize,
+        end: usize,
+    ) -> RopeSlice<'a> {
+        assert!(start <= end);
+        assert!(end <= node.text_info().bytes as usize);
+
+        // Early-out shortcut for taking a slice of the full thing.
+        if start == 0 && end == node.byte_count() {
+            if node.is_leaf() {
+                let text = node.leaf_text();
+                return RopeSlice(RSEnum::Light {
+                    text,
+                    char_count: count_chars(text) as Count,
+                    utf16_surrogate_count: count_utf16_surrogates(text) as Count,
+                    line_break_count: count_line_breaks(text) as Count,
+                });
+            } else {
+                return RopeSlice(RSEnum::Full {
+                    node,
+                    start_info: TextInfo {
+                        bytes: 0,
+                        chars: 0,
+                        utf16_surrogates: 0,
+                        line_breaks: 0,
+                    },
+                    end_info: TextInfo {
+                        bytes: node.byte_count() as Count,
+                        chars: node.char_count() as Count,
+                        utf16_surrogates: node.utf16_surrogate_count() as Count,
+                        line_breaks: node.line_break_count() as Count,
+                    },
+                });
+            }
+        }
+
+        // Find the deepest node that still contains the full range given.
+        let mut n_start = start;
+        let mut n_end = end;
+        let mut node = node;
+        'outer: loop {
+            match *(node as &Node) {
+                // Early out if we reach a leaf, because we can do the
+                // simpler lightweight slice then.
+                Node::Leaf(ref text) => {
+                    let start_byte = n_start;
+                    let end_byte = n_end;
+                    return RopeSlice(RSEnum::Light {
+                        text: &text[start_byte..end_byte],
+                        char_count: count_chars(&text[start_byte..end_byte]) as Count,
+                        utf16_surrogate_count: count_utf16_surrogates(&text[start_byte..end_byte])
+                            as Count,
+                        line_break_count: count_line_breaks(&text[start_byte..end_byte]) as Count,
```

Comment:
I kept the pattern here from `new_with_range` but the `&text[start_byte..end_byte]` can probably be extracted out